third_party/move/tools/move-lint/src/lint/rules/bool_comparison.rs

```diff
@@ -0,0 +1,157 @@
+                Some(func_name)
+            },
+            _ => None,
+        }
+    }
+
+    // This function examines the provided expression to identify if it contains a comparison
+    // between a boolean value and a variable (either a temporary or a local variable). If such
+    // a comparison is found, it generates a diagnostic message suggesting a more direct
+    // expression. For example, instead of `x == true`, it suggests using `x` directly, and
+    // for `x == false`, it suggests using `!x`. The diagnostic message is emitted as a warning
+    // to guide the user towards more idiomatic usage of boolean comparisons.
+    fn check_boolean_comparison(
+        &mut self,
+        cond: &ExpData,
+        func_env: &FunctionEnv,
+        env: &GlobalEnv,
+        diags: &mut Vec<Diagnostic<FileId>>,
+    ) {
+        if let ExpData::Call(_, oper, args) = &cond {
+            if let (Some(first_arg), Some(second_arg)) = (args.get(0), args.get(1)) {
+                if let (
+                    ExpData::Value(_, Value::Bool(a)),
+                    ExpData::Temporary(_, _) | ExpData::LocalVar(_, _),
+                ) = (first_arg.as_ref(), second_arg.as_ref())
+                {
+                    let var_name = self
+                        .get_var_name_or_func_name_from_exp(second_arg, func_env, env)
+                        .expect("Expected to get a variable name");
+
+                    let diagnostic_msg = match (oper, a) {
+                        (Operation::Eq, true) | (Operation::Neq, false) => Some(format!(
+                            "Use {} directly instead of comparing it to {}.",
+                            var_name, a
+                        )),
+                        (Operation::Eq, false) | (Operation::Neq, true) => Some(format!(
+                            "Use !{} directly instead of comparing it to {}.",
+                            var_name, a
+                        )),
+                        _ => None,
+                    };
+
+                    if let Some(diagnostic_msg) = diagnostic_msg {
+                        add_diagnostic_and_emit(
+                            &env.get_node_loc(cond.node_id()),
+                            &diagnostic_msg,
+                            codespan_reporting::diagnostic::Severity::Warning,
+                            env,
+                            diags,
+                        );
+                    }
+                } else {
+                    for exp in args {
+                        self.check_boolean_comparison(exp, func_env, env, diags);
+                    }
+                }
+            }
+
+            if let Some(exp_val) = args.get(1) {
+                if let ExpData::Value(_, Value::Bool(b)) = &exp_val.as_ref() {
+                    let var_name = self
+                        .get_var_name_or_func_name_from_exp(&args[0], func_env, env)
+                        .expect("Expected to get a variable name");
```

Comment:
Why can you expect to get a variable name here? We have not checked anything about `oper` or `args[0]` to be able to expect this. I believe the linter will panic upon seeing code that violates this expectation.

Reply:
panic is handled